In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import pandas as pd

# Load the dataset from the specified path
file_path = '/content/car data.csv'
df = pd.read_csv(file_path)

# Display the first 5 rows of the DataFrame
display(df.head())

,Car_Name,Year,Selling_Price,Present_Price,Driven_kms,Fuel_Type,Selling_type,Transmission,Owner
0,ritz,2014,3.35,5.59,27000,Petrol,Dealer,Manual,0
1,sx4,2013,4.75,9.54,43000,Diesel,Dealer,Manual,0
2,ciaz,2017,7.25,9.85,6900,Petrol,Dealer,Manual,0
3,wagon r,2011,2.85,4.15,5200,Petrol,Dealer,Manual,0
4,swift,2014,4.60,6.87,42450,Diesel,Dealer,Manual,0


### Understanding the Dataset Features

Let's first explore the general information about the dataset, including data types and non-null values. Then we'll look at descriptive statistics for numerical columns and unique values for categorical columns to get a complete overview of the features.

In [3]:
# Display basic information about the DataFrame
print("DataFrame Info:")
df.info()

DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 301 entries, 0 to 300
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Car_Name       301 non-null    object 
 1   Year           301 non-null    int64  
 2   Selling_Price  301 non-null    float64
 3   Present_Price  301 non-null    float64
 4   Driven_kms     301 non-null    int64  
 5   Fuel_Type      301 non-null    object 
 6   Selling_type   301 non-null    object 
 7   Transmission   301 non-null    object 
 8   Owner          301 non-null    int64  
dtypes: float64(2), int64(3), object(4)
memory usage: 21.3+ KB


In [4]:
# Display descriptive statistics for numerical columns
print("\nDescriptive Statistics for Numerical Features:")
display(df.describe())


Descriptive Statistics for Numerical Features:


,Year,Selling_Price,Present_Price,Driven_kms,Owner
count,301.000000,301.000000,301.000000,301.000000,301.000000
mean,2013.627907,4.661296,7.628472,36947.205980,0.043189
std,2.891554,5.082812,8.642584,38886.883882,0.247915
min,2003.000000,0.100000,0.320000,500.000000,0.000000
25%,2012.000000,0.900000,1.200000,15000.000000,0.000000
50%,2014.000000,3.600000,6.400000,32000.000000,0.000000
75%,2016.000000,6.000000,9.900000,48767.000000,0.000000
max,2018.000000,35.000000,92.600000,500000.000000,3.000000


In [5]:
# Display unique values and their counts for categorical columns
print("\nUnique Values for Categorical Features:")
for column in df.select_dtypes(include='object').columns:
    print(f"\nColumn '{column}':")
    print(df[column].value_counts())



Unique Values for Categorical Features:

Column 'Car_Name':
Car_Name
city                  26
corolla altis         16
verna                 14
fortuner              11
brio                  10
                      ..
Honda Activa 125       1
Hero Hunk              1
Hero  Ignitor Disc     1
Hero  CBZ Xtreme       1
Bajaj  ct 100          1
Name: count, Length: 98, dtype: int64

Column 'Fuel_Type':
Fuel_Type
Petrol    239
Diesel     60
CNG         2
Name: count, dtype: int64

Column 'Selling_type':
Selling_type
Dealer        195
Individual    106
Name: count, dtype: int64

Column 'Transmission':
Transmission
Manual       261
Automatic     40
Name: count, dtype: int64


Based on the initial exploration:

*   **Brand Goodwill**: The `Car_Name` column lists different car models, which could indirectly relate to brand goodwill. However, a direct 'goodwill' score is not present. This might require external data or feature engineering from existing `Car_Name` and pricing (`Present_Price`, `Selling_Price`) data.
*   **Horsepower**: This feature is **not directly available** in the dataset.
*   **Mileage**: The `Driven_kms` column provides information about the kilometers a car has been driven. If 'mileage' refers to fuel efficiency (e.g., km/liter or miles/gallon), that is **not directly available**, although `Fuel_Type` is present, which is a related factor.

### Extracting Car Brand from `Car_Name`

The `Car_Name` column contains various car models. We can try to extract the primary brand name from these entries to create a new feature that could be a proxy for 'brand goodwill'.

In [6]:
# Function to extract car brand from Car_Name
def extract_car_brand(car_name):
    # This is a simplified approach; more robust parsing might be needed for complex names.
    # Often, the first word is the brand, or a common brand name appears early.
    car_name = str(car_name).lower()
    if 'maruti' in car_name or 'swift' in car_name or 'ritz' in car_name or 'ciaz' in car_name or 'wagon r' in car_name or 'ertiga' in car_name or 'dzire' in car_name or 'celerio' in car_name or 'baleno' in car_name or 'alto' in car_name or 'ignis' in car_name or 's-cross' in car_name or 'sx4' in car_name:
        return 'Maruti'
    elif 'hyundai' in car_name or 'verna' in car_name or 'creta' in car_name or 'i20' in car_name or 'i10' in car_name or 'elantra' in car_name or 'eon' in car_name or 'venue' in car_name or 'santro' in car_name:
        return 'Hyundai'
    elif 'honda' in car_name or 'city' in car_name or 'jazz' in car_name or 'amaze' in car_name or 'wrv' in car_name or 'brio' in car_name or 'mobilio' in car_name:
        return 'Honda'
    elif 'toyota' in car_name or 'fortuner' in car_name or 'innova' in car_name or 'etios' in car_name or 'yaris' in car_name or 'glanza' in car_name or 'camry' in car_name or 'corolla' in car_name:
        return 'Toyota'
    elif 'mahindra' in car_name or 'scorpio' in car_name or 'xuv500' in car_name or 'bolero' in car_name or 'thar' in car_name or 'tuv300' in car_name:
        return 'Mahindra'
    elif 'ford' in car_name or 'ecosport' in car_name or 'figo' in car_name or 'endeavour' in car_name or 'freestyle' in car_name:
        return 'Ford'
    elif 'tata' in car_name or 'nexon' in car_name or 'tiago' in car_name or 'harrier' in car_name or 'altroz' in car_name or 'safari' in car_name or 'zest' in car_name:
        return 'Tata'
    elif 'chevrolet' in car_name or 'beat' in car_name or 'cruze' in car_name or 'enjoy' in car_name or 'tavera' in car_name:
        return 'Chevrolet'
    elif 'renault' in car_name or 'kwid' in car_name or 'duster' in car_name or 'triber' in car_name:
        return 'Renault'
    elif 'volkswagen' in car_name or 'polo' in car_name or 'vento' in car_name or 'ameo' in car_name:
        return 'Volkswagen'
    elif 'audi' in car_name or 'q3' in car_name or 'a4' in car_name or 'q7' in car_name:
        return 'Audi'
    elif 'bmw' in car_name or 'x1' in car_name or '3 series' in car_name:
        return 'BMW'
    elif 'mercedes' in car_name or 'benz' in car_name or 'c-class' in car_name:
        return 'Mercedes-Benz'
    elif 'skoda' in car_name or 'rapid' in car_name or 'octavia' in car_name or 'superb' in car_name:
        return 'Skoda'
    elif 'jeep' in car_name or 'compass' in car_name:
        return 'Jeep'
    elif 'mitsubishi' in car_name or 'pajero' in car_name or 'outlander' in car_name:
        return 'Mitsubishi'
    elif 'isuzu' in car_name or 'd-max' in car_name:
        return 'Isuzu'
    elif 'land rover' in car_name or 'range rover' in car_name:
        return 'Land Rover'
    elif 'force' in car_name:
        return 'Force Motors'
    elif 'datsun' in car_name:
        return 'Datsun'
    elif 'fiat' in car_name:
        return 'Fiat'
    elif 'royal enfield' in car_name:
        return 'Royal Enfield'
    elif 'hero' in car_name:
        return 'Hero MotoCorp'
    elif 'bajaj' in car_name:
        return 'Bajaj Auto'
    elif 'tv' in car_name or 'tvs' in car_name:
        return 'TVS'
    elif 'yamaha' in car_name:
        return 'Yamaha'
    elif 'ktm' in car_name:
        return 'KTM'
    elif 'vespa' in car_name:
        return 'Vespa'
    elif 'suzuki' in car_name and not 'maruti' in car_name: # Handle Suzuki cars that are not Maruti badged
        return 'Suzuki'
    else:
        return 'Other'

# Apply the function to create a new 'Brand' column
df['Brand'] = df['Car_Name'].apply(extract_car_brand)

# Display the top 10 most common brands and the first few rows with the new column
print("\nTop 10 Car Brands Extracted:")
print(df['Brand'].value_counts().head(10))

print("\nDataFrame with new 'Brand' column:")
display(df.head())


Top 10 Car Brands Extracted:
Brand
Honda            69
Toyota           49
Hyundai          47
Maruti           46
Bajaj Auto       25
Royal Enfield    17
Other            13
Hero MotoCorp    13
TVS               8
Yamaha            8
Name: count, dtype: int64

DataFrame with new 'Brand' column:


,Car_Name,Year,Selling_Price,Present_Price,Driven_kms,Fuel_Type,Selling_type,Transmission,Owner,Brand
0,ritz,2014,3.35,5.59,27000,Petrol,Dealer,Manual,0,Maruti
1,sx4,2013,4.75,9.54,43000,Diesel,Dealer,Manual,0,Maruti
2,ciaz,2017,7.25,9.85,6900,Petrol,Dealer,Manual,0,Maruti
3,wagon r,2011,2.85,4.15,5200,Petrol,Dealer,Manual,0,Maruti
4,swift,2014,4.60,6.87,42450,Diesel,Dealer,Manual,0,Maruti


### Feature Engineering and Preprocessing

Before training a regression model, we need to preprocess the data. This involves:
1.  **Creating 'Age' feature**: Deriving the age of the car from the `Year` column.
2.  **One-Hot Encoding**: Converting categorical features into numerical format.
3.  **Dropping irrelevant columns**: Removing original `Year` and `Car_Name` columns.

In [7]:
# Calculate current year for 'Age' feature
current_year = 2024 # Assuming current year is 2024

# Create 'Age' feature
df['Age'] = current_year - df['Year']

# Drop the original 'Year' and 'Car_Name' columns as they are no longer needed
df = df.drop(columns=['Year', 'Car_Name'])

print("DataFrame after creating 'Age' and dropping 'Year' and 'Car_Name':")
display(df.head())

DataFrame after creating 'Age' and dropping 'Year' and 'Car_Name':


,Selling_Price,Present_Price,Driven_kms,Fuel_Type,Selling_type,Transmission,Owner,Brand,Age
0,3.35,5.59,27000,Petrol,Dealer,Manual,0,Maruti,10
1,4.75,9.54,43000,Diesel,Dealer,Manual,0,Maruti,11
2,7.25,9.85,6900,Petrol,Dealer,Manual,0,Maruti,7
3,2.85,4.15,5200,Petrol,Dealer,Manual,0,Maruti,13
4,4.60,6.87,42450,Diesel,Dealer,Manual,0,Maruti,10


In [8]:
# Identify categorical columns for one-hot encoding
categorical_cols = ['Fuel_Type', 'Selling_type', 'Transmission', 'Brand']

# Apply one-hot encoding
df = pd.get_dummies(df, columns=categorical_cols, drop_first=True)

print("\nDataFrame after One-Hot Encoding:")
display(df.head())
print(f"\nNew number of columns: {df.shape[1]}")


DataFrame after One-Hot Encoding:


,Selling_Price,Present_Price,Driven_kms,Owner,Age,Fuel_Type_Diesel,Fuel_Type_Petrol,Selling_type_Individual,Transmission_Manual,Brand_Hero MotoCorp,...,Brand_Hyundai,Brand_KTM,Brand_Mahindra,Brand_Maruti,Brand_Other,Brand_Royal Enfield,Brand_Suzuki,Brand_TVS,Brand_Toyota,Brand_Yamaha
0,3.35,5.59,27000,0,10,False,True,False,True,False,...,False,False,False,True,False,False,False,False,False,False
1,4.75,9.54,43000,0,11,True,False,False,True,False,...,False,False,False,True,False,False,False,False,False,False
2,7.25,9.85,6900,0,7,False,True,False,True,False,...,False,False,False,True,False,False,False,False,False,False
3,2.85,4.15,5200,0,13,False,True,False,True,False,...,False,False,False,True,False,False,False,False,False,False
4,4.60,6.87,42450,0,10,True,False,False,True,False,...,False,False,False,True,False,False,False,False,False,False



New number of columns: 21


### Training and Evaluating a Regression Model

Now, we will train a regression model to predict the car's `Selling_Price` based on the engineered and preprocessed features. We'll use a `RandomForestRegressor` for this task, as it's a robust and generally effective model for such predictions. We will split the data into training and testing sets to evaluate the model's performance on unseen data.

In [9]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import r2_score, mean_absolute_error

# Define features (X) and target (y)
X = df.drop('Selling_Price', axis=1)
y = df['Selling_Price']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print(f"Training set size: {X_train.shape[0]} samples")
print(f"Test set size: {X_test.shape[0]} samples")

# Initialize and train the Random Forest Regressor model
model = RandomForestRegressor(n_estimators=100, random_state=42)
model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = model.predict(X_test)

# Evaluate the model
r2 = r2_score(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)

print(f"\nModel Evaluation:")
print(f"R-squared (R2) score: {r2:.2f}")
print(f"Mean Absolute Error (MAE): {mae:.2f}")

Training set size: 240 samples
Test set size: 61 samples

Model Evaluation:
R-squared (R2) score: 0.96
Mean Absolute Error (MAE): 0.61


Conclusion:

The trained RandomForestRegressor model has demonstrated excellent performance in predicting car selling prices. With an R-squared (R2) score of 0.96, the model can explain 96% of the variance in the selling prices, indicating a very strong fit to the data. The Mean Absolute Error (MAE) of 0.61 suggests that, on average, the model's predictions are off by approximately 0.61 units (e.g., Lakhs if the Selling_Price is in Lakhs). This low MAE, combined with the high R2 score, implies that the model is highly accurate and can be reliably used for predicting car prices based on the features available in this dataset. This level of accuracy makes the model suitable for real-world applications in car dealerships, online marketplaces, or for individuals looking to buy or sell used cars, enabling them to estimate fair prices.